ocaml/xapi/xapi_vm_migrate.ml

```diff
@@ -216,12 +217,76 @@ let assert_sr_support_operations ~__context ~vdi_map ~remote ~local_ops
          op_supported_on_dest_sr sr remote_ops sm_record remote
      )
 
+(** [get_vdi_type vdi_ref vdi_format_map] returns the vdi type found in the 
+    [vdi_format_map] mapping for a given [vdi_ref]. If no type is found None
+    is returned. *)
+let get_vdi_type ~vdi_ref ~vdi_format_map =
+  List.assoc_opt vdi_ref vdi_format_map
+
+(** [assert_vdi_format_is_supported] checks that all VDIs in [vdi_map] are included in the list of 
+    supported image format of their corresponding SM. The type of the VDI is found in [vdi_format_map].
+    - If no VDI type is specified we just returned so no error is raised.
+    - If an SM reports an empty list of supported formats, we cannot verify compatibility and no error
+      is raised. So if the format is not actually supported, the failure will be detected later when
+      attempting to create the VDI using that image format. *)
+let assert_vdi_format_is_supported ~__context ~vdi_map ~vdi_format_map =
+  List.iter
+    (fun (vdi_ref, sr_ref) ->
+      let vdi_uuid = Db.VDI.get_uuid ~__context ~self:vdi_ref in
+      let sr_uuid = Db.SR.get_uuid ~__context ~self:sr_ref in
+      match List.assoc_opt vdi_ref vdi_format_map with
+      | None ->
+          debug "GTNDEBUG: read vdi %s, sr %s. No type specified for the VDI"
+            vdi_uuid sr_uuid
+      | Some ty -> (
+          (* To get the supported image format from SM we need the SR type because both have
+             the same type. *)
+          let sr_type = Db.SR.get_type ~__context ~self:sr_ref in
+          let sm_refs =
+            Db.SM.get_refs_where ~__context
+              ~expr:(Eq (Field "type", Literal sr_type))
+          in
+          match sm_refs with
+          | [sm_ref] ->
+              debug "GTNDEBUG: read vdi %s, sr %s. Type is %s" vdi_uuid sr_uuid
+                ty ;
+              let sm_formats =
+                Db.SM.get_supported_image_formats ~__context ~self:sm_ref
+              in
+              if sm_formats <> [] && not (List.mem ty sm_formats) then
+                raise
+                  Api_errors.(
+                    Server_error
+                      ( vdi_incompatible_type
```

Comment:
Currently we return incompatible type that makes sense I think but we can also create a new kind of error.